### DISCUSSION TITLE
Benchmark results slow with python3

### DISCUSSION THREAD

**Discussion by Holundermann** [seaweedfs/seaweedfs#2620] [Q&A]
Hello,
I am currently evaluating seaweedFS as object store for our projects. We mostly using python3 for our applications - therefore I wrote a small benchmark tool in python3 - nothing fancy.

I create a data set of # files (for example 10000) between 5 KB and 50 KB then I write it to the database and read it again. I tried my own implementation using 'requests' for the HTTP communication and I am using 'concurrent.futures.ProcessPoolExecutor' to spawn processes which do the read/write tasks.

With this implementation I am not even close to the internal benchmark one can run with weed - and I wonder how this is possible. I mean - python3 may not result in the fastest code - but so far I thought if the task is I/O bound - which I expect in this case - it should not have an big impact.

Test resulsts from weed - 1000 files with 50 KB: 
`docker run -it --net seaweed-evaluate_default --rm registry.my.com/dockerhub/chrislusf/seaweedfs benchmark -master=master:9333 -size=51200 -n 10000
This is SeaweedFS version 30GB 2.87 e185d90d linux amd64

------------ Writing Benchmark ----------
Completed 4168 of 10000 requests, 41.7% 4167.6/s 203.6MB/s
Completed 9132 of 10000 requests, 91.3% 4964.4/s 242.5MB/s

Concurrency Level:      16
Time taken for tests:   2.151 seconds
Complete requests:      10000
Failed requests:        0
Total transferred:      512314678 bytes
Requests per second:    4648.16 [#/sec]
Transfer rate:          232550.71 [Kbytes/sec]

Connection Times (ms)
              min      avg        max      std
Total:        0.7      3.4       219.1      8.5

Percentage of the requests served within a certain time (ms)
   50%      2.7 ms
   66%      3.4 ms
   75%      3.8 ms
   80%      4.1 ms
   90%      5.1 ms
   95%      6.0 ms
   98%      7.3 ms
   99%      8.5 ms
  100%    219.1 ms

------------ Randomly Reading Benchmark ----------
Completed 9225 of 10000 requests, 92.2% 9225.1/s 450.7MB/s

Concurrency Level:      16
Time taken for tests:   1.090 seconds
Complete requests:      10000
Failed requests:        0
Total transferred:      512316441 bytes
Requests per second:    9176.51 [#/sec]
Transfer rate:          459109.12 [Kbytes/sec]

Connection Times (ms)
              min      avg        max      std
Total:        0.2      1.6       19.3      1.4

Percentage of the requests served within a certain time (ms)
   50%      1.2 ms
   66%      1.8 ms
   75%      2.2 ms
   80%      2.5 ms
   90%      3.3 ms
   95%      4.1 ms
   98%      5.1 ms
   99%      5.9 ms
  100%     19.3 ms
`
My test results - 10000 files between 5 KB and 50 KB:
`
Prepared 10000 testfiles between 5.0KiB and 50.0KiB, 269.5MiB total.
Starting Seaweed write test...
Seaweed write test finished in 7.02 seconds. 1425.26 files/second - 38.4MiB per second.
Starting Seaweed read test...
Seaweed read test finished in 7.11 seconds. 1406.21 files/second - 37.9MiB per second.
`
And this is the code responsible for writing the data:
`
session = requests.Session()

def post_data(files: Dict[str, Any]) -> None:
    master_response = session.get("{}/dir/assign?collection=benchmark_1".format(seaweed_master))
    master_responses.append(master_response)
    public_url = master_response.json()["publicUrl"]
    fid = master_response.json()["fid"]
    volume_write_responses.append(
        session.post("{}/{}".format(public_url, fid), files=files)
    )
`

Am I not using seaweed correct?

**Comment by chrislusf** [comment]
Reusing connections matters.

One difference is that the Go code uses a consistent gRPC connection to assign file ids.
In other places, the HTTP connections are re-used.

Not sure whether Python reuse the http connections.

**Reply by Holundermann** [reply]
I will try to adapt my code and ensure I reuse my connections. Further I will check out how I can make a persistend gRPC connection to the master. Is it somewhere documented how to use gRPC to assign file ids?

**Reply by chrislusf** [reply]
Not really any documentation but just proto file https://github.com/chrislusf/seaweedfs/blob/master/weed/pb/master.proto#L16

It is just standard gRPC. Usually just the proto is needed, and you should be able to figure out how to use it in your own language.
